rmf_rviz_plugin/src/RmfPanel.cpp

```diff
@@ -13,404 +13,758 @@
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_finish = value;
-  Q_EMIT configChanged();
+  rviz_common::Panel::save(config);
 }
 
-void RmfPanel::set_loop_robot(const QString& value)
+// Q_SLOTS
+// Actions
+void RmfPanel::queue_delivery() 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_robot = value;
+  std::string start = _start_waypoint_selector->currentText().toStdString();
+  std::string end = _end_waypoint_selector->currentText().toStdString();
+  Delivery delivery;
+
+  // If either start or end are empty, the delivery should probably not be queued
+  if (start.empty() || end.empty())
+  {
+    RCLCPP_INFO(_node->get_logger(), "Waypoint input is empty string; Delivery not queued in plan.");
+    return;
+  }
+
+  delivery.task_id = generate_task_uuid(3);
+  delivery.pickup_place_name = start;
+  delivery.dropoff_place_name = end;
+
+  auto delivery_time = _time_selector->time();
+
+  int insertPos = 0;
+  for (auto it = _queued_deliveries.begin(); it != _queued_deliveries.end();
+       it++) {
+    if (delivery_time <= it->first) 
+    {
+      break;
+    } else {
+      insertPos++;
+    }
+  }
+
+  std::pair<QTime, Delivery> data =
+      std::pair<QTime, Delivery>(delivery_time, delivery);
+  _queued_deliveries.insert(_queued_deliveries.begin() + insertPos, data);
+
+  RCLCPP_INFO(_node->get_logger(), "Queued delivery request");
   Q_EMIT configChanged();
 }
 
-void RmfPanel::set_loop_num(const QString& value)
+void RmfPanel::queue_loop() 
 {
-  std::lock_guard<std::mutex> lock(_mutex);
-  _loop_num = value;
+  std::string start = _start_waypoint_selector->currentText().toStdString();
+  std::string end = _end_waypoint_selector->currentText().toStdString();
+  Loop loop;
+
+  // If either start or end are empty, the delivery should probably not be queued
+  if (start.empty() || end.empty())
+  {
+    RCLCPP_INFO(_node->get_logger(), "Waypoint input is empty string; Delivery not queued in plan.");
+    return;
+  }
+
+  loop.task_id = generate_task_uuid(4);
+  loop.robot_type = _fleet_selector->currentText().toStdString();
+  loop.num_loops = _repeat_count_selector->value();
+  loop.start_name = start;
+  loop.finish_name = end;
+
+  auto loop_time = _time_selector->time();
+
+  int insertPos = 0;
+  for (auto it = _queued_loops.begin(); it != _queued_loops.end(); it++) 
+  {
+    if (loop_time <= it->first) 
+    {
+      break;
+    } else 
+    {
+      insertPos++;
+    }
+  }
+
+  std::pair<QTime, Loop> data = std::pair<QTime, Loop>(loop_time, loop);
+  _queued_loops.insert(_queued_loops.begin() + insertPos, data);
+
+  RCLCPP_INFO(_node->get_logger(), "Queued loop request");
   Q_EMIT configChanged();
 }
 
-// Delivery update functions 
-void RmfPanel::update_delivery_task_id()
+void RmfPanel::pop_delivery() 
 {
-  set_delivery_task_id(_delivery_task_id_editor->text());
+  auto msg = _queued_deliveries.begin()->second;
+  std::lock_guard<std::mutex> lock(_mutex);
```

Comment:
`pop_delivery()` and `pop_loop()` seems to only be call serially in `pop_plan()`, is there a reason to have a lock here and on line 431?